### PR TITLE
some minor additions

### PR DESCRIPTION
**Pull request by in0finite**

### CODE REVIEW

File: src/DemoFile/DemoFileReader.Parallel.cs [hunk 1]
```diff
@@ -91,10 +95,10 @@ await reader.MoveNextCoreAsync(cmd.Command, cmd.IsCompressed, cmd.Size, cancella
             {
             }
 
-            return new[] {result};
+            return new[] {initialResult, result};
```
Comment (in0finite) [inline src/DemoFile/DemoFileReader.Parallel.cs:98]: it seems that `initialResult` is missing

File: src/DemoFile/DemoParser.cs [hunk 2]
```diff
@@ -62,7 +67,7 @@ protected DemoParser()
     public ref BaseUserMessageEvents BaseUserMessageEvents => ref _baseUserMessageEvents;
     public ref TempEntityEvents TempEntityEvents => ref _tempEntityEvents;
 
-    public CDemoFileHeader FileHeader { get; private set; } = new();
+    public CDemoFileHeader? FileHeader { get; private set; }
```
Comment (in0finite) [inline src/DemoFile/DemoParser.cs:70]: it doesn't make sense to initialize this to invalid values. Also, the message has not been read yet.

File: src/DemoFile/DemoParser.Entities.cs [hunk 3]
```diff
@@ -85,6 +85,16 @@ protected T GetCachedSingletonEntity<T>(ref CHandle<T, TGameParser> handle) wher
...
     }
 
+    /// <summary>
+    /// Fastest way to check if entity handle exists.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public bool EntityHandleExists(CEntityIndex entityIndex, uint serialNumber)
+    {
+        var e = _entities[entityIndex.Value];
+        return e != null && e.SerialNumber == serialNumber;
```
[the remark this answers is not in the document]
Comment (in0finite) [reply src/DemoFile/DemoParser.Entities.cs:95]: I don't have any benchmarks, but consider following:

- `GetEntityByHandle()` requires `CHandle` which can only be calculated from `CEntityIndex ` and `serialNumber`, which includes a lot of bit-wise arithmetics
- it then extracts  `CEntityIndex ` and `serialNumber` from `CHandle` , which also does bit-wise arithmetics
- it also does type checking

So, based on this, it seems obvious that new overload is faster.

File: src/DemoFile/DemoParser.cs [hunk 4]
```diff
@@ -24,6 +24,11 @@ public abstract partial class DemoParser<TGameParser>
     /// </summary>
     public Action? OnCommandFinish;
 
+    /// <summary>
+    /// Event fired when the current demo command has finished.
+    /// </summary>
+    public Action? OnCommandFinishPersistent;
```
[the remark this answers is not in the document]
Comment (in0finite) [reply src/DemoFile/DemoParser.cs:30]: Whatever you want, I just need equivalent functionality, this is very useful for parallel parsing where you can't do MoveNextAsync() and then process data

File: src/DemoFile/DemoParser.Entities.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (saul) [reply src/DemoFile/DemoParser.Entities.cs:95]: Yes but that's trivially optimisable by the JIT. This library is heavily optimised, any performance related change needs to be benchmarked

[the remark this answers is not in the document]
Comment (in0finite) [reply src/DemoFile/DemoParser.Entities.cs:95]: Okay, I will remove the comment saying it's the fastest method. Would you be willing then to add it ? It doesn't break anything.

File: src/DemoFile/DemoFileReader.Parallel.cs (hunk 1 above)
Comment (in0finite) [reply src/DemoFile/DemoFileReader.Parallel.cs:98]: I couldn't find which test needs to be modified. Do you want me to add new test for this case ?

File: src/DemoFile/DemoParser.Entities.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (saul) [reply src/DemoFile/DemoParser.Entities.cs:95]: If it is much faster then it makes sense to keep. If it's not faster then it serves no purpose

[the remark this answers is not in the document]
Comment (in0finite) [reply src/DemoFile/DemoParser.Entities.cs]: I removed it. I will use reflection to get access to entities array.

File: src/DemoFile/DemoFileReader.FullPacket.cs [hunk 5]
```diff
@@ -45,6 +45,7 @@ internal readonly record struct FullPacketRecord(
     private int _fullPacketTickOffset;
 
     internal IReadOnlyList<FullPacketRecord> FullPackets => _fullPackets;
+    public int NumFullPackets => _fullPackets.Count;
```
Comment (saul) [inline src/DemoFile/DemoFileReader.FullPacket.cs:48]:
```suggestion
    public int FullPacketCount => _fullPackets.Count;
```

File: src/DemoFile/DemoFileReader.Parallel.cs (hunk 1 above)
Comment (saul) [reply src/DemoFile/DemoFileReader.Parallel.cs:120]: Yes please - once that's done this PR is good to merge

**Pull request merged**